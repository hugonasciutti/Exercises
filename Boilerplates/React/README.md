# Features

Dry boiler plate with:

- Webpack dev server
- React
- Babel Env
