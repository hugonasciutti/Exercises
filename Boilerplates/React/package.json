{
  "name": "boilerplatereact",
  "version": "1.0.0",
  "description": "",
  "main": "index.js",
  "scripts": {
    "start": "webpack-dev-server"
  },
  "author": "",
  "license": "ISC",
  "dependencies": {
    "final-form": "^4.5.2",
    "react": "^16.3.2",
    "react-dom": "^16.3.2"
  },
  "devDependencies": {
    "babel-core": "^6.26.0",
    "babel-loader": "^7.1.4",
    "babel-preset-env": "^1.6.1",
    "babel-preset-react": "^6.24.1",
    "babel-preset-stage-0": "^6.24.1",
    "html-webpack-plugin": "^3.2.0",
    "webpack": "^4.6.0",
    "webpack-cli": "^2.0.15",
    "webpack-dev-server": "^3.1.3"
  }
}
