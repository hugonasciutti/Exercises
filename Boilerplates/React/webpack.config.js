const HtmlWebpackPlugin = require('html-webpack-plugin')

module.exports = {
  mode: 'development',

  entry: './src/index.jsx',

  output: {
    path: `${__dirname}/dist`,
    filename: 'app.js'
  },

  devtool: 'source-map',

  module: {
    rules: [
      { test: /\.(js|jsx)?$/, loader: 'babel-loader', exclude: /node_modules/ }
    ]
  },
  plugins: [
    new HtmlWebpackPlugin({
      title: 'Boilerplate',
      template: './src/index.ejs'
    })
  ]
}
