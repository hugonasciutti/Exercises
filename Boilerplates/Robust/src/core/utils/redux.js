export const createReducer = (initialState, actions) =>
  (state = initialState, action) => {
    const reducer = actions[action.type]
    return reducer ? reducer(state, action) : state
  }

export default createReducer
