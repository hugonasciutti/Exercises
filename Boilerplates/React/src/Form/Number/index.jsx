import React, { Component } from 'react'
import PropTypes from 'prop-types'
import isEmpty from 'lodash/isEmpty'
import numbro from 'core/utils/numbro'

class Number extends Component {
  constructor(props) {
    super(props)

    this.state = {
      value: this.getFormatValue(props.value)
    }
    this.onBlur = this.onBlur.bind(this)
  }

  onChangeInput({ value }) {
    const { regexKeydown } = this.props

    value = value.replace(regexKeydown(), '')

    this.setState({ value })

    this.props.onChange(isEmpty(value) ? value : this.getUnformatValue(value))
  }

  onBlur() {
    const { value } = this.state

    !isEmpty(value) && this.setState({ value: this.getFormatValue(value) })
  }

  getFormatValue(value) {
    const { format } = this.props

    return isEmpty(value) ? '' : numbro(value).format(format)
  }

  getUnformatValue(value) {
    return numbro().unformat(this.getFormatValue(value))
  }

  render() {
    const { name, required, placeholder } = this.props
    const { value } = this.state

    return (
      <input
        className="rw-number form-control"
        type="text"
        name={name}
        value={value}
        onChange={({ target }) => this.onChangeInput(target)}
        onBlur={this.onBlur}
        placeholder={placeholder}
        required={required}
      />
    )
  }
}

Number.propTypes = {
  name: PropTypes.string,
  value: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.number
  ]),
  onChange: PropTypes.func.isRequired,
  placeholder: PropTypes.string,
  required: PropTypes.bool.isRequired,
  format: PropTypes.string.isRequired,
  regexKeydown: PropTypes.func.isRequired
}

Number.defaultProps = {
  required: false,
  format: numbro.cultureData().formats.fullWithTwoDecimalsNoCurrency,
  regexKeydown: () => /[^0-9.,]/g
}

export default Number
