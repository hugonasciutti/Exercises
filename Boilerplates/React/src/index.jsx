import React, { Component } from 'react'
import { render } from 'react-dom'
import Form from './Form'

class App extends Component {
  render() {
    return (
      <div>
        <Input />
      </div>
    )
  }
}

render(<App />, document.getElementById('app'))