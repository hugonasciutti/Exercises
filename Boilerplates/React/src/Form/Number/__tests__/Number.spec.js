import React from 'react'
import { shallow, mount } from 'enzyme'
import { buildEvent } from 'core/unit-test/utils'
import Number from '../'

describe('Number Component', () => {
  it('should render', () => {
    const component = shallow(<Number onChange={jest.fn()} />)

    expect(component).toMatchSnapshot()
  })

  it('should render replace default props', () => {
    const component = shallow(<Number name="a" value={1} placeholder="a" onChange={jest.fn()} required />)

    expect(component).toMatchSnapshot()
  })

  it('should call onChange', () => {
    const onChange = jest.fn()
    const value = 1500.50
    const component = shallow(<Number onChange={onChange} />)

    component.simulate('change', buildEvent('1500.50'))

    expect(onChange).toHaveBeenCalledWith(value)
  })

  it('should clean input value', () => {
    const onChange = jest.fn()
    const component = shallow(<Number value={1} onChange={onChange} />)

    component.simulate('change', buildEvent(''))

    expect(onChange).toHaveBeenCalledWith('')
  })

  it('should enter letters in default regex formart', () => {
    const onChange = jest.fn()
    const value = 1500.50
    const component = shallow(<Number onChange={onChange} />)

    component.simulate('change', buildEvent('15bb00.50a'))

    expect(onChange).toHaveBeenCalledWith(value)
  })

  it('should call format on blur', () => {
    const onChange = jest.fn()
    const component = mount(<Number onChange={onChange} />)

    component.state().value = '1500'
    component.simulate('blur')

    expect(component.state().value).toBe('1,500.00')
  })
})
