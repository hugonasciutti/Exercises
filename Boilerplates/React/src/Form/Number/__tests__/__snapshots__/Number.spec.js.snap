// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Number Component should render 1`] = `
<input
  className="rw-number form-control"
  onBlur={[Function]}
  onChange={[Function]}
  required={false}
  type="text"
  value=""
/>
`;

exports[`Number Component should render replace default props 1`] = `
<input
  className="rw-number form-control"
  name="a"
  onBlur={[Function]}
  onChange={[Function]}
  placeholder="a"
  required={true}
  type="text"
  value=""
/>
`;
